File: 백준/Gold/15573. 채굴/채굴.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int N, M, K;
    static int[][] map;
    static int[] dy = {-1, 1, 0, 0}, dx = {0, 0, -1, 1};
    static Queue<int[]> q = new ArrayDeque<>();
    static boolean[][] v;
    public static void main(String[] args) throws IOException {
//        System.setIn(Files.newInputStream(Paths.get("src/input.txt")));
//        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());
        map = new int[N][M];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int low = 1, high = 1_000_000;
        while (low < high) {
            int d = (low + high) >> 1;

            if (BFS(d) < K) {
                low = d + 1;
            } else{
                high = d;
            }
        }
        bw.write(high + "");
        bw.flush();
        bw.close();
        br.close();
    }

    static int BFS(int d){
        prepareMining(d);
        int cnt = q.size();

        while (!q.isEmpty()) {
            int[] cur = q.poll();

            for (int i = 0; i < 4; i++) {
                int ny = dy[i] + cur[0], nx = dx[i] + cur[1];
                if (ny >= N || nx >= M || ny < 0 || nx < 0 || v[ny][nx]) {
                    continue;
                }
                v[ny][nx] = true;
                if (map[ny][nx] > d) {
                    continue;
                }
                q.add(new int[]{ny, nx});
                cnt += 1;
            }
        }
        return cnt;
    }

    static void prepareMining(int d){
        v = new boolean[N][M];

        for (int i = 0; i < N; i++) {
            for (int j : new int[]{0, M - 1}) {
                v[i][j] = true;
                if(map[i][j] > d){
                    continue;
                }
                q.add(new int[]{i, j});
            }
        }

        for (int i = 1; i < M - 1; i++) {
            v[0][i] = true;
            if(map[0][i] > d){
                continue;
            }
            q.add(new int[]{0, i});
        }
    }
}

File: 백준/Gold/15573. 채굴/채굴.py
import sys
from collections import deque

# print = sys.stdout.write
input = sys.stdin.readline

# 반드시 아래 두 라인 주석처리 후 업로드
# f = open("input.txt", "rt")
# input = f.readline
# 위 두라인 주석처리 후 업로드

DIRECTIONS = ((-1, 0), (1, 0), (0, -1), (0, 1))

N, M, K = map(int, input().split())
maps = [list(map(int, input().split())) for _ in range(N)]


def main():
    start, end = 1, 10**6

    while start < end:
        d = (start + end) >> 1

        if BFS(d) < K:
            start = d + 1
        else:
            end = d

    print(end)


def prepare_mining(d, v):
    q = deque()
    for i in range(N):
        for j in (0, M - 1):
            if maps[i][j] <= d:
                q.append((i, j))
                v[i][j] = True

    for i in range(1, M - 1):
        if maps[0][i] <= d:
            q.append((0, i))
            v[0][i] = True
    return q


def BFS(d):
    v = [[False] * M for _ in range(N)]
    q = prepare_mining(d, v)
    cnt = len(q)

    while q:
        y, x = q.popleft()

        for dy, dx in DIRECTIONS:
            ny, nx = dy + y, dx + x
            if ny >= N or nx >= M or ny < 0 or nx < 0 or v[ny][nx] or maps[ny][nx] > d:
                continue
            q.append((ny, nx))
            v[ny][nx] = True
            cnt += 1

    return cnt


if __name__ == "__main__":
    main()
